app/src/main/java/info/nightscout/androidaps/plugins/general/openhumans/OpenHumansUploader.kt

```
@@ -0,0 +1,549 @@
+        if (oAuthTokens != null && this.isEnabled(PluginType.GENERAL)) {
+            try {
+                val jsonObject = JSONObject()
+                jsonObject.put("structureVersion", structureVersion)
+                jsonObject.put("queuedOn", System.currentTimeMillis())
+                generator(jsonObject)
+                val queueItem = OHQueueItem(
+                    file = file,
+                    content = jsonObject.toString()
+                )
+                MainApp.getDbHelper().createOrUpdate(queueItem)
+                RxBus.send(OpenHumansFragment.UpdateQueueEvent)
+            } catch (e: JSONException) {
+                e.printStackTrace()
+            }
+        }
+    }
+
+    fun login(authCode: String): Completable =
+        openHumansAPI.exchangeAuthToken(authCode)
+            .doOnSuccess {
+                oAuthTokens = it
+            }
+            .flatMap { openHumansAPI.getProjectMemberId(it.accessToken) }
+            .doOnSuccess {
+                projectMemberId = it
+                copyExistingDataToQueue()
+                RxBus.send(OpenHumansFragment.UpdateViewEvent)
+            }
+            .ignoreElement()
+
+    fun logout() {
+        cancelWorker()
+        copyDisposable?.dispose()
+        isSetup = false
+        oAuthTokens = null
+        projectMemberId = null
+        MainApp.getDbHelper().clearOpenHumansQueue()
+        RxBus.send(OpenHumansFragment.UpdateViewEvent)
+    }
+
+    private fun copyExistingDataToQueue() {
+        copyDisposable?.dispose()
+        var currentProgress = 0L
+        var maxProgress = 0L
+        val increaseCounter = {
+            currentProgress++
+            //Updating the notification for every item drastically slows down the operation
+            if (currentProgress % 1000L == 0L) showOngoingNotification(maxProgress, currentProgress)
+        }
+        copyDisposable = Completable.fromCallable { MainApp.getDbHelper().clearOpenHumansQueue() }
```

If the copyDisposable gets disposed, can't that cause an UndeliverableException if `fromCallable` emits?

---

DatabaseHelper already swallows all neccessary exceptions and returns an empty list in case of an error.